aimmo-game/simulation/turn_manager.py

```diff
@@ -1,81 +1,123 @@
 import logging
 import requests
-import threading
 import time
+from Queue import PriorityQueue
 from threading import Lock
-from simulation.action import ACTIONS
+from threading import Thread
+
+from simulation.action import PRIORITIES
 
 LOGGER = logging.getLogger(__name__)
 
 
-class WorldStateProvider:
+class GameStateProvider:
     """
     Thread-safe container for the world state.
 
     TODO: think about changing to snapshot rather than lock?
     """
 
     def __init__(self):
-        self._world_state = None
+        self._game_state = None
         self._lock = Lock()
 
-    def lock_and_get_world(self):
+    def __enter__(self):
         self._lock.acquire()
-        return self._world_state
+        return self._game_state
 
-    def release_lock(self):
+    def __exit__(self, type, value, traceback):
         self._lock.release()
 
-    def set_world(self, new_world_state):
+    def set_world(self, new_game_state):
         self._lock.acquire()
-        self._world_state = new_world_state
+        self._game_state = new_game_state
         self._lock.release()
 
-world_state_provider = WorldStateProvider()
+
+state_provider = GameStateProvider()
 
 
-class TurnManager(threading.Thread):
+class TurnManager(Thread):
     """
     Game loop
     """
     daemon = True
 
-    def __init__(self, game_state, end_turn_callback):
-        world_state_provider.set_world(game_state)
+    def __init__(self, game_state, end_turn_callback, concurrent_turns=True):
+        state_provider.set_world(game_state)
         self.end_turn_callback = end_turn_callback
+        self.concurrent_turns = concurrent_turns
         super(TurnManager, self).__init__()
 
-    def _update_environment(self, game_state):
-        num_avatars = len(game_state.avatar_manager.active_avatars)
-        game_state.world_map.reconstruct_interactive_state(num_avatars)
-
     def run_turn(self):
-        try:
-            game_state = world_state_provider.lock_and_get_world()
-
-            for avatar in game_state.avatar_manager.active_avatars:
-                turn_state = game_state.get_state_for(avatar)
-                try:
-                    data = requests.post(avatar.worker_url, json=turn_state).json()
-                except ValueError as err:
-                    LOGGER.info("Failed to get turn result: %s", err)
+        if self.concurrent_turns:
+            self.run_concurrent_turn()
+        else:
+            self.run_sequential_turn()
+
+    def run_sequential_turn(self):
+        '''
+        Get and apply each avatar's action in turn.
+        '''
+        with state_provider as game_state:
+            avatars = game_state.avatar_manager.active_avatars
```

Comment:
Do we want to release the lock here? We immediatly get it back inside [_register_action](https://github.com/ocadotechnology/aimmo/pull/127/files#diff-0df92163d570b5e36adea6e4166f2f9aR107) and also the loop. We could end up giving different states to different avatars.

Slightly off-topic: I'm not sure what other thread uses the state_provider lock? WorkerManager has its own lock. Could we have race conditions between TurnManager and WorkerManager (eg if an avatar is removed during a turn)?